applications/OpenBox.md

```diff
@@ -0,0 +1,459 @@
+    * NFT:         seller    -> buyer
+    * value:       buyer     -> seller
+    * listingFee:  contract  -> marketowner
+    */
+    function createMarketSale(
+    address nftContract,
+    uint256 id
+    ) public payable nonReentrant {
+
+    MarketItem storage item = marketItems[id]; //should use storge!!!!
+    uint price = item.price;
+    uint tokenId = item.tokenId;
+
+    require(msg.value == price, "Please submit the asking price");
+    require(IERC721(nftContract).getApproved(tokenId) == address(this), "NFT must be approved to market");
+
+    IERC721(nftContract).transferFrom(item.seller, msg.sender, tokenId);
+
+    payable(marketowner).transfer(listingFee);
+    item.seller.transfer(msg.value);
+
+    item.buyer = payable(msg.sender);
+    item.state = State.Release;
+    _itemSoldCounter.increment();    
+
+    emit MarketItemSold(
+      id,
+      nftContract,
+      tokenId,
+      item.seller,
+      msg.sender,
+      price,
+      State.Release
+      );    
+      }
+
+     /**
+     * Returns all unsold market items
+     * condition: 
+     *  1) state == Created
+     *  2) buyer = 0x0
+     *  3) still have approve
+     */
+     function fetchActiveItems() public view returns (MarketItem[] memory) {
+     return fetchHepler(FetchOperator.ActiveItems);
+     }
+
+ 
+     function fetchMyPurchasedItems() public view returns (MarketItem[] memory) {
+     return fetchHepler(FetchOperator.MyPurchasedItems);
+     }
+
+     function fetchMyCreatedItems() public view returns (MarketItem[] memory) {
+     return fetchHepler(FetchOperator.MyCreatedItems);
+     }
+
+     enum FetchOperator { ActiveItems, MyPurchasedItems, MyCreatedItems}
+
+     function fetchHepler(FetchOperator _op) private view returns (MarketItem[] memory) {     
+     uint total = _itemCounter.current();
+
+     uint itemCount = 0;
+     for (uint i = 1; i <= total; i++) {
+      if (isCondition(marketItems[i], _op)) {
+        itemCount ++;
+       }
+      }
+
+     uint index = 0;
+     MarketItem[] memory items = new MarketItem[](itemCount);
+     for (uint i = 1; i <= total; i++) {
+      if (isCondition(marketItems[i], _op)) {
+        items[index] = marketItems[i];
+        index ++;
+        }
+      }
+     return items;
+    } 
+
+    function isCondition(MarketItem memory item, FetchOperator _op) private view returns (bool){
+    if(_op == FetchOperator.MyCreatedItems){ 
+      return 
+        (item.seller == msg.sender
+          && item.state != State.Inactive
+        )? true
+         : false;
+     }else if(_op == FetchOperator.MyPurchasedItems){
+      return
+        (item.buyer ==  msg.sender) ? true: false;
+     }else if(_op == FetchOperator.ActiveItems){
+      return 
+        (item.buyer == address(0) 
+          && item.state == State.Created
+          && (IERC721(item.nftContract).getApproved(item.tokenId) == address(this))
+        )? true
+         : false;
+      }else{
+        return false;
+      }
+      }
+
+      }
```

Comment:
What are these showing?

Reply:
This part is a contract based on Solidity's NFT MarketPlace